test/unit/BPool/BPool_SwapExactAmountIn.t.sol

```diff
@@ -0,0 +1,139 @@
+  }
+
+  function test_RevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.swapExactAmountIn(tokenIn, tokenAmountIn, tokenOut, expectedAmountOut, spotPriceAfterSwap);
+  }
+
+  function test_RevertWhen_PoolIsNotFinalized() external {
+    bPool.set__finalized(false);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_PoolNotFinalized.selector);
+    bPool.swapExactAmountIn(tokenIn, tokenAmountIn, tokenOut, expectedAmountOut, spotPriceAfterSwap);
+  }
+
+  function test_RevertWhen_TokenInIsNotBound() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    bPool.swapExactAmountIn(makeAddr('unknown token'), tokenAmountIn, tokenOut, expectedAmountOut, spotPriceAfterSwap);
+  }
+
+  function test_RevertWhen_TokenOutIsNotBound() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenNotBound.selector);
+    bPool.swapExactAmountIn(tokenIn, tokenAmountIn, makeAddr('unknown token'), expectedAmountOut, spotPriceAfterSwap);
+  }
+
+  function test_RevertWhen_TokenAmountInExceedsMaxAllowedRatio(uint256 tokenAmountIn_) external {
+    tokenAmountIn_ = bound(tokenAmountIn_, bmul(tokenInBalance, MAX_IN_RATIO) + 1, type(uint256).max);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenAmountInAboveMaxRatio.selector);
+    bPool.swapExactAmountIn(tokenIn, tokenAmountIn_, tokenOut, 0, 0);
+  }
+
+  function test_RevertWhen_SpotPriceBeforeSwapExceedsMaxPrice() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_SpotPriceAboveMaxPrice.selector);
+    bPool.swapExactAmountIn(tokenIn, tokenAmountIn, tokenOut, expectedAmountOut, spotPriceBeforeSwap - 1);
+  }
+
+  function test_RevertWhen_CalculatedTokenAmountOutIsLessThanMinAmountOut() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_TokenAmountOutBelowMinOut.selector);
+    bPool.swapExactAmountIn(tokenIn, tokenAmountIn, tokenOut, expectedAmountOut + 1, spotPriceAfterSwap);
+  }
+
+  function test_RevertWhen_SpotPriceAfterSwapExceedsSpotPriceBeforeSwap() external {
+    // it should revert
+    // skipping since the code for this is unreachable without manually
+    // overriding `calcSpotPrice` in a mock:
+    // P_{sb} = \frac{\frac{b_i}{w_i}}{\frac{b_o}{w_o}}
+    // P_{sa} = \frac{\frac{b_i + a_i}{w_i}}{\frac{b_o - a_o}{w_o}}
+    // ...and both a_i (amount in) and a_o (amount out) are uints
+    vm.skip(true);
+  }
+
+  function test_RevertWhen_SpotPriceAfterSwapExceedsMaxPrice() external {
+    // it should revert
+    vm.expectRevert(IBPool.BPool_SpotPriceAboveMaxPrice.selector);
+    bPool.swapExactAmountIn(tokenIn, tokenAmountIn, tokenOut, expectedAmountOut, spotPriceAfterSwap - 1);
+  }
+
+  function test_RevertWhen_SpotPriceBeforeSwapExceedsTokenRatioAfterSwap() external {
+    // it should revert
+    vm.skip(true);
```

Comment:
have a set of values that triggers this :)

```suggestion
    uint256 tokenAmountIn_ = 2983194183788398023993031849354427183591957587;
    uint256 balanceTokenIn_ = 3683096585543175594085153402580232440657855508718480656945;
    uint256 weightTokenIn_ = 11341738085061414652;
    uint256 balanceTokenOut_ = 1819895046949535393784309338278140699348077743733566505520;
    uint256 weightTokenOut_ = 11429833665193825207;
    uint256 swapFee_ = 19438253123623515;

    vm.mockCall(tokenIn, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(balanceTokenIn_)));
    vm.mockCall(tokenOut, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(balanceTokenOut_)));
    bPool.set__records(tokenIn, IBPool.Record({bound: true, index: 0, denorm: weightTokenIn_}));
    bPool.set__records(tokenOut, IBPool.Record({bound: true, index: 1, denorm: weightTokenOut_}));
    bPool.set__swapFee(swapFee_);

    // it should revert
    vm.expectRevert(IBPool.BPool_SpotPriceBeforeAboveTokenRatio.selector);
    bPool.swapExactAmountIn(tokenIn, tokenAmountIn_, tokenOut, 0, type(uint256).max);
```

Reply:
how did you generate this? a fuzz run?

Reply:
i used the fuzzer of legacy `swapExactAmountIn` tests to generate a counterproof, to hit the specific revert a combination of `testFail` and `vm.expectRevert` can be used